### PR TITLE
Allow parsing of non-root productions

### PR DESCRIPTION
**Pull request by mccolljr**
This is currently a work in progress, but this PR will allow parsing non-root productions using a parser. This can be useful when implementing `Parseable` or using `ParseTypeWith`.

### CODE REVIEW

**Comment by mccolljr** [comment]
@alecthomas wanted to solicit initial feedback before I finish up this PR using this appraoch

File: _examples/go.mod [hunk 1]
```diff
@@ -5,14 +5,11 @@ go 1.18
 require (
 	github.com/alecthomas/assert/v2 v2.0.3
 	github.com/alecthomas/go-thrift v0.0.0-20170109061633-7914173639b2
-	github.com/alecthomas/kong v0.5.0
-	github.com/alecthomas/participle/v2 v2.0.0-alpha1
+	github.com/alecthomas/kong v0.6.1
```
Comment (mccolljr) [inline _examples/go.mod:8]: Wanted to get all the example tests passing to make sure I didn't break anything, so updated the go.mod file and the `thrift` test.

**Comment by alecthomas** [comment]
Looks pretty reasonable to me, thanks!

**Review by alecthomas — APPROVED** [review]

**Comment by alecthomas** [comment]
Thanks!

**Comment by alecthomas** [comment]
I just merged 1261c4d90338839f363ac13d6252c50b0b942f75 which makes the parser itself generic. As part of that I added a `func ParserForProduction[P, G any](parser *Parser[G]) (*Parser[P], error)` which safely retrieves a parser for a production.

**Pull request merged**